mono/mini/interp/interp.c

```diff
@@ -5032,60 +5025,50 @@ call:;
 
 			/*
 			 * First arg is the object.
+			 * a constructor returns void, but we need to return the object we created
 			 */
 			if (m_class_is_valuetype (newobj_class)) {
 				MonoType *t = m_class_get_byval_arg (newobj_class);
-				memset (&valuetype_this, 0, sizeof (stackval));
 				if (!m_class_is_enumtype (newobj_class) && (t->type == MONO_TYPE_VALUETYPE || (t->type == MONO_TYPE_GENERICINST && mono_type_generic_inst_is_valuetype (t)))) {
-					sp->data.p = vt_sp;
-					valuetype_this.data.p = vt_sp;
+					sp [0].data.p = vt_sp; // return value
+					sp [1].data.p = vt_sp; // first parameter
 				} else {
-					sp->data.p = &valuetype_this;
+					memset (sp, 0, sizeof (*sp));
+					sp [1].data.p = &sp [0].data; // first parameter is return value
```

Comment:
If stackval were to have extra data at the start, it is not clear if this should be `&sp [0].data]` or `&sp [0]`.
It is presently moot though.